dll/shellext/fontext/CFontExt.cpp

```diff
@@ -363,3 +365,176 @@ STDMETHODIMP CFontExt::GetClassID(CLSID *lpClassId)
+    }
+
+    BOOL bOK = TRUE;
+    CAtlArray<CStringW> FontPaths;
+    HDROP hDrop = (HDROP)GlobalLock(medium.hGlobal);
+    if (hDrop)
+    {
+        WCHAR szPath[MAX_PATH];
+        UINT iFile, cFiles = DragQueryFile(hDrop, 0xFFFFFFFF, NULL, 0);
+        for (iFile = 0; iFile < cFiles; ++iFile)
+        {
+            DragQueryFile(hDrop, iFile, szPath, _countof(szPath));
+
+            if (PathIsDirectoryW(szPath))
+            {
+                bOK = FALSE;
+                break;
+            }
+
+            LPCWSTR pchDotExt = PathFindExtensionW(szPath);
+            if (_wcsicmp(pchDotExt, L".ttf") != 0 &&
+                _wcsicmp(pchDotExt, L".ttc") != 0 &&
+                _wcsicmp(pchDotExt, L".otf") != 0 &&
+                _wcsicmp(pchDotExt, L".otc") != 0 &&
+                _wcsicmp(pchDotExt, L".fon") != 0 &&
+                _wcsicmp(pchDotExt, L".fnt") != 0)
+            {
+                bOK = FALSE;
+                break;
+            }
+
+            FontPaths.Add(szPath);
+        }
+
+        GlobalUnlock(medium.hGlobal);
+    }
+    ReleaseStgMedium(&medium);
+
+    if (!bOK)
+        return E_FAIL;
+
+    WCHAR szFontsDir[MAX_PATH];
+    GetWindowsDirectoryW(szFontsDir, _countof(szFontsDir));
+    PathAppendW(szFontsDir, L"Fonts");
+
+    HKEY hkeyFonts = NULL;
+    RegOpenKeyExW(HKEY_LOCAL_MACHINE,
+                  L"SOFTWARE\\Microsoft\\Windows NT\\CurrentVersion\\Fonts",
+                  0,
+                  KEY_READ | KEY_WRITE,
+                  &hkeyFonts);
+    if (!hkeyFonts)
+        return E_FAIL;
+
+    for (size_t iItem = 0; iItem < FontPaths.GetCount(); ++iItem)
+    {
+        HRESULT hr = DoInstallFontFile(FontPaths[iItem], szFontsDir, hkeyFonts);
+        if (FAILED(hr))
+        {
+            bOK = FALSE;
+            break;
+        }
+    }
+
+    RegCloseKey(hkeyFonts);
+
+    if (bOK)
+    {
+        SendMessageW(HWND_BROADCAST, WM_FONTCHANGE, 0, 0);
+    }
+
+    // TODO: Show message
+
+    return bOK ? S_OK : E_FAIL;
+}
+
+HRESULT CFontExt::DoGetFormatETC(IDataObject* pDataObj, FORMATETC& formatETC)
+{
+    if (!pDataObj)
+        return E_POINTER;
+
+    formatETC.cfFormat = CF_HDROP;
+    formatETC.ptd = NULL;
+    formatETC.dwAspect = DVASPECT_CONTENT;
+    formatETC.lindex = -1;
+    formatETC.tymed = TYMED_HGLOBAL;
+
+    HRESULT hr = pDataObj->QueryGetData(&formatETC);
+    return hr;
+}
+
+HRESULT CFontExt::DoInstallFontFile(LPCWSTR pszFontPath, LPCWSTR pszFontsDir, HKEY hkeyFonts)
+{
+    WCHAR szDestFile[MAX_PATH];
+    LPCWSTR pszFileTitle = PathFindFileName(pszFontPath);
+
+    WCHAR szFontName[512];
+    if (!DoGetFontTitle(pszFontPath, szFontName))
+        return E_FAIL;
+
+    RemoveFontResourceW(pszFileTitle);
+
+    StringCchCopyW(szDestFile, sizeof(szDestFile), pszFontsDir);
+    PathAppendW(szDestFile, pszFileTitle);
+    if (!CopyFileW(pszFontPath, szDestFile, FALSE))
+        return E_FAIL;
+
+    if (!AddFontResourceW(pszFileTitle))
```

Comment:
Why is this needed?

Reply:
https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-addfontresourcew
>The AddFontResource function adds the font resource from the specified file to the system font table. The font can subsequently be used for text output by any application.

`CopyFileW` just copys the file to the filesystem. `CopyFileW` won't install the font file.

Reply:
Font engine became registry-based, so simply copying doesn't install font.

Reply:
Ah yes, this makes it available immediately to the applications without a reboot.